zcash_client_sqlite/src/lib.rs

```diff
@@ -558,6 +558,59 @@ impl<'a, P: consensus::Parameters> WalletWrite for DataConnStmtCache<'a, P> {
         })
     }
 
+    #[cfg(feature = "unstable")]
+    fn remove_tx(&mut self, txid: &TxId) -> Result<(), Self::Error> {
+        self.transactionally(|up| {
+            // Find the transaction to be deleted.
+            let tx_ref = up.stmt_select_tx_ref(txid)?;
+
+            // Delete witnesses for received notes that will be deleted.
+            up.wallet_db.conn.execute_named(
+                "DELETE FROM sapling_witnesses WHERE note IN (
+                    SELECT rn.id_note
+                    FROM received_notes rn
+                    WHERE rn.tx = :tx
+                )",
+                &[(":tx", &tx_ref)],
+            )?;
```

Comment:
This code was not present in the upstream Android SDK logic, but was necessary to enable deleting just-mined transactions (so I could use the test harness). Note however that omitting this logic would not have prevented the bug from being triggered, as old witnesses get periodically pruned, so old-enough transactions would not run into any foreign key issue.

Reply:
Another idea: would it make sense to join to the `blocks` table and only delete transactions where they aren't associated with any block?

Reply:
If we want this to specifically be `remove_unmined_tx` then yes that should work. That's not how the Android SDK logic was implemented, but based on how it was _used_ I think that should be sufficient. @ccjernigan?

Reply:
I've now implemented this change.